utils/python/CIME/XML/env_mach_specific.py

```diff
@@ -189,7 +194,16 @@ def _compute_actions(self, nodes, child_tag, compiler, debug, mpilib):
                 for child in node:
                     expect(child.tag == child_tag, "Expected %s element" % child_tag)
                     if (self._match_attribs(child.attrib, compiler, debug, mpilib)):
-                        result.append( (child.get("name"), child.text) )
+                        val = child.text
+                        if val is not None:
+                            # We allow a couple special substitutions for these fields
+                            for repl_this, repl_with in [("$COMPILER", compiler), ("$MPILIB", mpilib)]:
+                                val = val.replace(repl_this, repl_with)
+
+                            val = self.get_resolved_value(val)
+                            expect("$" not in val, "Not safe to leave unresolved items in env var value: '%s'" % val)
```

Comment:
There is at least one case of a place where you need to leave environment variables unresolved - on mira there are variables which must only be resolved after the job has booted on the compute node, I believe that that system would fail with this expect.